Added a button to take user to top when page is scrolled down

share-with-me opened the pull request:
Created a button that lets the user scroll up. It is hidden in mobile and tablet devices.

sushain97 commented:
Hm, if I quickly scroll up and down (a couple times) and then stay down, the arrow fades back and forth :) I think you need to cancel some events or use CSS transitions? Unsure. But I'm relatively certain that's not how you want it to to behave.

The UI looks great (and the moving up action)! I'll take a closer look at the code after the build errors have been fixed and the other issue I mention is looked at.

share-with-me commented:
@sushain97 I have fixed the flickering of button. It was basically because of `slow` effect of Jquery animation that got queued and got executed even after we scroll up and down so fast. I have fixed that and updated PR. Here is the [link](https://drive.google.com/open?id=0BwrSYlOsMCyWLTNsQUVDWXpLczg)

share-with-me commented on assets/js/util.js:
```diff
@@ -248,3 +248,30 @@ function synchronizeTextareaHeights() {
+    $(window).scroll(function () {
+        if($(window).scrollTop() > amountScrolled) {
+            $('#go-top').fadeIn('fast');
+        } 
```
What do we mean by trailing spaces?

sushain97 commented on index.html.in:
```diff
@@ -531,5 +531,6 @@
             <div class="lead text-warning well well-sm" data-text="Enable_JS_Warning"></div>
             <noscript><p><img src="@include_piwik_url@/piwik.php?idsite=@include_piwik_siteid@" style="border: 0;" alt=""></p></noscript>
         </div>
+        <button class="btn btn-primary" id = "go-top"><i class="fa fa-arrow-up" aria-hidden="true"></i></button>
```
Extra spacing: `id = "go-top"`. Also, use `#gotoTop` or `#backToTop` for the sake of consistency.

sushain97 commented on assets/js/util.js:
```diff
@@ -248,3 +248,30 @@ function synchronizeTextareaHeights() {
+//  is run on desktop or larger screens
+
+/*  Following is the function that provides functionality of go-to-top button */
+$(function () {
```
This should go inside the `$(document).ready(function () {` already in `utils.js`.

sushain97 commented on assets/js/util.js:
```diff
@@ -248,3 +248,30 @@ function synchronizeTextareaHeights() {
+    if($(window).width() < TEXTAREA_AUTO_RESIZE_MINIMUM_WIDTH) {
+        return;
+    }
+    var amountScrolled = 300; /*:: amount by which user must scroll to view the button */
```
Add these variables at the top (~L10).

sushain97 commented on assets/css/style.css:
```diff
@@ -321,3 +321,21 @@ html[dir=rtl] .pull-right:not(#rtlChecker) {
+    opacity: 1;
+}
+
+.fa-arrow-up {
```
Instead of styling all of them, use the ID, .e.g. `#gotoTop .fa-arrow-up`.

sushain97 commented on assets/css/style.css:
```diff
@@ -321,3 +321,21 @@ html[dir=rtl] .pull-right:not(#rtlChecker) {
 }
+
+/* Styling for the go-to-top button */
+#go-top{
```
Space between brace and identifier.

sushain97 commented on assets/js/util.js:
```diff
@@ -248,3 +248,30 @@ function synchronizeTextareaHeights() {
 /*:: import {iso639Codes, iso639CodesInverse} from "./localization.js" */
+
+//  As the dynamic extending of boxes is disabled in devices with width < 768, button appears when the application
+//  is run on desktop or larger screens
```
This can be established by a CSS class, I believe. Take a look at Bootstrap's documentation.

share-with-me commented:
@sushain97 I have updated the changes but I am unable to test them as my `make` command is failing (this issue #121 ). Once I test it, I will update the commits here for you to review.

share-with-me commented:
@sushain97 I have made the changes as requested but I made a force push as I had not merged the changes from the current master onto my remote on this branch. Would keep that in mind henceforth.

sushain97 commented:
@share-with-me hm I'm not sure why force push says that. However, don't worry about it. I will squash your commits when I merge them in.

sushain97 commented on assets/css/style.css:
```diff
@@ -321,3 +321,21 @@ html[dir=rtl] .pull-right:not(#rtlChecker) {
     float: right !important;
 }
+
+/* Styling for the back-to-top button */
```
Move these styles to the `General` section and remove this comment.

sushain97 commented on assets/css/style.css:
```diff
@@ -321,3 +321,21 @@ html[dir=rtl] .pull-right:not(#rtlChecker) {
+    opacity: 1;
+}
+
+.fa-arrow-up {
```
Use something like `#backToTop .fa-arrow-up`. This needs to not affect all the arrows.

sushain97 commented on assets/js/util.js:
```diff
@@ -9,6 +9,9 @@ var SPACE_KEY_CODE = 32, ENTER_KEY_CODE = 13,
 var TEXTAREA_AUTO_RESIZE_MINIMUM_WIDTH = 768;
 
+var amountScrolled = 300; /*:: amount by which user must scroll to view the button */
+var transitionTime = 600; /*:: time taken for the transition to move webpage to top */
```
Move these to be part of the var above, use the same variable naming style. No need for comments if the variable names are good.

share-with-me replied:
@sushain97 by same naming style, do we mean I need to change to something like:
var AMOUNT_SCROLLED = 300;
var TRANSITION_TIME = 600;??

> Move these to be part of the var above,

Means I have to write these above the TEXT_AUTO_RESIZE_MINIMUM_WIDTH right?
I would remove the variable names.

sushain97 commented on assets/js/util.js:
```diff
@@ -119,6 +122,28 @@ $(document).ready(function () {
         $('a[data-target=#' + $(this).attr('id') + ']').parents('li').removeClass('active');
     });
+
+    /*  Following is the function that provides functionality of back-to-top button */
```
There's no need for this comment :) We don't want to clutter the file.

sushain97 commented on assets/js/util.js:
```diff
@@ -119,6 +122,28 @@ $(document).ready(function () {
+    $('#backToTop').click(function () {
+        $('html, body').animate({
+            scrollTop: 0
+        }, transitionTime);
```
I think `'fast'` is sufficient here.

share-with-me replied:
@sushain97  I will remove the var `transitionTime` from the code and just put a `fast` there!

sushain97 commented on assets/js/util.js:
```diff
@@ -119,6 +122,28 @@ $(document).ready(function () {
+        return false;
+    });
+
+    /* Function for backToTop ends here */
```
Same goes for this comment. Please remove.

share-with-me commented:
@sushain97 I would not force push henceforth. Really sorry for the inconvenience. And yep, on changes and will update this PR today. Thanks once again!

share-with-me commented:
@sushain97 I have made the changes. Please have a look! Thanks!

sushain97 commented on assets/js/analyzer.js:
```diff
@@ -143,6 +143,10 @@ function populateSecondaryAnalyzerList() {
 }
 
 function analyze() {
+    if($('#primaryAnalyzerMode').val() === null) {
```
Why are these changes here?

share-with-me replied:
@sushain97  Even I was confused. I did a git diff on my local machine between this branch and master branch and the only changes are with respect to the functionality of button. Could you help me understand why it might be happening ? I am updating this PR to check if this is resolved. Thanks!

sushain97 commented on assets/js/util.js:
```diff
@@ -6,6 +6,7 @@
 var SPACE_KEY_CODE = 32, ENTER_KEY_CODE = 13,
     HTTP_OK_CODE = 200, HTTP_BAD_REQUEST_CODE = 400,
     XHR_LOADING = 3, XHR_DONE = 4;
+var AMOUNT_SCROLLED = 300;
```
This needs a more descriptive name.

share-with-me replied:
@sushain97 `AMOUNT_SCROLLED_ON_WEBPAGE` would work?

sushain97 commented on .gitignore:
```diff
@@ -12,3 +12,5 @@ assets/js/config.js
 .Trashes
 Thumbs.db
 ehthumbs.db
+.DS_Store
```
These are already ignored in master's gitignore. Rebase your commits on master?

share-with-me replied:
modified my .gitignore file itself.

sushain97 commented on assets/js/util.js:
```diff
@@ -6,6 +6,7 @@
 var SPACE_KEY_CODE = 32, ENTER_KEY_CODE = 13,
     HTTP_OK_CODE = 200, HTTP_BAD_REQUEST_CODE = 400,
     XHR_LOADING = 3, XHR_DONE = 4;
+var AMOUNT_SCROLLED_ON_WEBPAGE = 300;
```
How about, `BACK_TO_TOP_ACTIVATION_HEIGHT`? Try combining it with the `var` on L11 as well.

share-with-me replied:
On it!

sushain97 commented:
@jonorthwash what do you think of this feature?

Get it using:

    git checkout -b share-with-me-mono-apertium master
    git pull https://github.com/share-with-me/apertium-html-tools.git mono-apertium

Aside from changing the button from being blue colored to normal colored, I think it's good to go. Thoughts?

share-with-me commented:
@sushain97 meanwhile, I will update this PR with default button :D

share-with-me commented:
umm, @sushain97 , what color do we mean by normal colored? Should I go with default or basic?

share-with-me commented:
@sushain97 umm, what is the status of this PR? What further changes should I make? I really look forward to closing this :D .

jonorthwash commented:
@share-with-me, I was going to take a look at this, which I haven't done yet.  I'll plan to take a look tomorrow, but if you don't hear from me further about it in a day or two, poke me.

share-with-me commented:
@jonorthwash awesome! meanwhile I will work on finishing other existing PRs :D .

share-with-me commented:
@jonorthwash Umm, you asked me to remind you to look into this PR. Thanks!

share-with-me commented:
@jonorthwash um, did you look at the PR? Thanks!

jonorthwash commented:
About to do it now, sorry about that.

jonorthwash commented:
I like the feature, but I have some questions about it:
1. Why doesn't it appear as soon as the buttons at the top go out of view?  And by "buttons at the top", I mean everything, including the language dropdown.
2. Why is it hidden on mobile?
3. How is it hidden on mobile?  Even when I "request desktop site", my WQHD mobile device doesn't give the button in either orientation.
4. Why is the right margin of the button different from other content on the page?
5. Is it a problem that the button covers up page content (although it is transparent...).

share-with-me commented:
@jonorthwash  Thank you so much for reviewing :D.

> Why doesn't it appear as soon as the buttons at the top go out of view? And by "buttons at the top", I mean everything, including the language dropdown.

Currently, the height by which we scroll down the webpage for button to appear is static ie 300. I have not taken into consideration the visibility of other UI components in viewport to trigger the visibility of button.
1. Should I embed it that way? (ie, the button would appear in case the top UI elements go out of viewport?) **OR**
2. should I modify this height to say 100 (or other), so that as soon as the user scrolls down even by just a little, he can view the button?

> Why is it hidden on mobile?

1. The functionality this feature was designed to address was, when the content in the original and destination textareas reaches beyond a limit, the button should appear. 
2. The current JS allows dynamic re-rendering of textareas (with increases height) depending on the content length in only desktop devices(ie col-md-* devices)  thereby disabling the y-scroll. However, in the mobile devices, we are not allowing this. The mobile devices show a y-scroll if the content length increases.
3. **I still believe we can (and IMHO, should) include the button in mobile devices too. This would let the user scroll to top in any device.**

> How is it hidden on mobile? Even when I "request desktop site", my WQHD mobile device doesn't give the button in either orientation.

1. I have used (col-xs-) and (col-sm-) hidden to achieve the functionality. The CSS senses the device width and renders the CSS accordingly. Therefore, I believe, even when 'request for desktop' site is made, the device width being that of mobile/tablet, it does not render the UI element.
2. I am inclined towards enabling the button in mobile and tablet devices too, as I mentioned above.

> Why is the right margin of the button different from other content on the page?

1. If I give it the same margin as the other content, it would overlap the content for some part of scroll on the webpage.
2. Is the current right-margin correct? Any changes required in the position of the button?

> Is it a problem that the button covers up page content (although it is transparent...)?

1. Could you be more specific by page content? Do you mean the translation contents? Do we mean something like the following? (The interface appears this way on decreasing the window width. The button overlaps the translation box contents).

![screen shot 2017-04-17 at 12 05 16 pm](https://cloud.githubusercontent.com/assets/8894636/25081142/2559a6ae-2366-11e7-8c68-0aba65c04576.png)

2. However, on normal width window, the screen appears like this.

![screen shot 2017-04-17 at 12 07 02 pm](https://cloud.githubusercontent.com/assets/8894636/25081172/631a061e-2366-11e7-8a7f-b69e9facce3d.png)

> Btw, it's currently live at http://turkic.apertium.org

Feels awesome to see the feature live :D. 

Other than the above questions, a few questions:
1. Is the color of button good? Should I change it to the brown-ish color as that of all other btn-default UI components?
2. Is the positioning of button correct?
3. Is the opacity and transparency correct?

IMHO, we should include this feature in mobile and tablet devices too.

sushain97 commented:
> However, on normal width window, the screen appears like this.

There is no such thing as a normal width, really.

> Is the color of button good? Should I change it to the brown-ish color as that of all other btn-default UI components?

Use the other color. The blue color is only for primary buttons.

> IMHO, we should include this feature in mobile and tablet devices too.
 
No, this feature should not be enabled in mobile. It will cover far too much of the screen and doesn't add much value because we disable the textbox resizing/growth feature on mobile. It's also really easy to scroll to the top on phones since scrolling has variable velocity. Making the button smaller would make it too small of a target = bad UX.

share-with-me commented:
> There is no such thing as a normal width, really.

Agreed! My bad! What I meant was the entire screen size the window takes in my laptop device by default, when I maximise it.

> Use the other color. The blue color is only for primary buttons.

On it.

> No, this feature should not be enabled in mobile. It will cover far too much of the screen and doesn't add much value because we disable the textbox resizing/growth feature on mobile. 

It will indeed take up a fair amount of screen space. 

> It's also really easy to scroll to the top on phones since scrolling has variable velocity.

This point is highly convincing to not enable this feature on the mobile devices. Agreed :D.

> Making the button smaller would make it too small of a target = bad UX.

Yep, definitely not a good option.

share-with-me commented:
@sushain97 I have updated the PR. Please have a look! Thanks!

share-with-me commented:
@sushain97  @jonorthwash  @xavivars , I have updated this PR with the requested features. Please have a look and let me know if any changes are required! Thanks!

jonorthwash commented:
> Is the current right-margin correct? Any changes required in the position of the button?

It's a little weird.

>> Is it a problem that the button covers up page content (although it is transparent...)?
>
> Could you be more specific by page content? Do you mean the translation contents? 

Yes.

> Is the opacity and transparency correct?

Semi-transparent might be nice?

share-with-me commented:
> It's a little weird.

Um, what could be a suitable position?

> Semi-transparent might be nice?

Would decrease the opacity!

share-with-me commented:
@jonorthwash @sushain97 Tried with a semi-transparent touch! Please have a look!

sushain97 commented:
Looks good to me! If there are no objections from @jonorthwash or @xavivars I will squash and merge this PR.

sushain97 commented on assets/js/util.js:
```diff
@@ -119,6 +119,25 @@ $(document).ready(function () {
+    $(window).scroll(function () {
+        if($(window).scrollTop() > BACK_TO_TOP_ACTIVATION_HEIGHT) {
+            $('#backToTop').removeClass('hidden-xs hidden-sm hidden-md hidden-lg');
+            $('#backToTop').addClass('hidden-xs hidden-sm visible-md visible-lg');
```
Use method chaining: https://schier.co/blog/2013/11/14/method-chaining-in-javascript.html

sushain97 replied:
Why are you removing then adding the same class? This happens below as well.

share-with-me replied:
While removing the class, I remove `hidden-xs hidden-sm **hidden-md hidden-lg**` (basically all the hidden classes as we have to make them visible) and while adding the class, I add `hidden-xs hidden-sm  **visible-md visible-lg** `. The `hidden-xs hidden-sm` classes stay constant as this feature should not be rendered on mobile devices. Thus, when the scroll passes the height, I add the `visible-md visible-lg` classes to ensure they are visible only in `*-md *-lg` screens and are hidden in `*-sm *-xs` devices.

sushain97 replied:
Yes, `hidden-xs` and `hidden-sm` stay constant. So, my question is: why are you removing them and immediately adding them back? Why not simply never touch them?

share-with-me replied:
Um, if we do not explicitly mention `hidden-xs hidden-sm`, would it appear in the small screen devices too? Because, otherwise, there is no place where I explicitly mention that this feature should be hidden on mobile! Maybe I am wrong!

sushain97 replied:
Basically, what I'm saying is that your code

```
$('#backToTop').removeClass('hidden-xs hidden-sm hidden-md hidden-lg');
$('#backToTop').addClass('hidden-xs hidden-sm visible-md visible-lg');
```

== 

```
$('#backToTop').removeClass('hidden-md hidden-lg');
$('#backToTop').addClass('hidden-xs hidden-sm visible-md visible-lg');
```

Does that make sense?

On the other hand, if you intend for a class to always be applied to a particular element, that should happen in the HTML, not the JS.

share-with-me replied:
Um, yep these do make sense! I have modified the PR and added `hidden-xs hidden-sm` to the HTML element directly. I only fiddle around with the `visible/hidden-md,lg` in the JS.
Thanks @sushain97  for this! It definitely makes a lot more sense to me now.

sushain97 commented on assets/js/util.js:
```diff
@@ -119,6 +119,25 @@ $(document).ready(function () {
+        }
+        else {
+            $('#backToTop').removeClass('hidden-xs hidden-sm visible-md visible-lg');
+            $('#backToTop').addClass('hidden-xs hidden-sm hidden-md hidden-lg');
```
Use method chaining here as well.

share-with-me commented:
@sushain97 I have added the method chaining! Please have a look.

sushain97 commented on assets/js/util.js:
```diff
@@ -119,6 +119,27 @@ $(document).ready(function () {
+    $(window).scroll(function () {
+        if($(window).scrollTop() > BACK_TO_TOP_ACTIVATION_HEIGHT) {
+            $('#backToTop')
+            .removeClass('hidden-xs hidden-sm hidden-md hidden-lg')
```
Huh, I'm surprised that TravisCI/Hound, i.e. ESLint doesn't catch this. I might have my rules configured incorrectly. Regardless, L126 and L127 here should be indented (just as 131 and 132).

share-with-me replied:
Ohh! I would indent it.

sushain97 commented on assets/js/util.js:
```diff
@@ -119,6 +119,27 @@ $(document).ready(function () {
+
+    $(window).scroll(function () {
+        if($(window).scrollTop() > BACK_TO_TOP_ACTIVATION_HEIGHT) {
+            $('#backToTop')
```
Looking at this again, would it not be simpler to do something like

```js
$('#backToTop').css('display', $(window).scrollTop() > BACK_TO_TOP_ACTIVATION_HEIGHT ? 'block' : 'initial')
````

share-with-me replied:
@sushain97 , Um, I tried optimising the code with both of the above methods and it works fine with both! However, when I add a `hidden-xs hidden-sm` class to the `button`, and run the above code, the functionality works fine but I am unable to hide the `button` on `-md -lg` devices even with a CSS of `display: none;`. Does it mean that adding `hidden-xs hidden-sm` implicitly apply `visible-md visible-lg`?

share-with-me replied:
```
$( function() {
    $('#backToTop').addClass('hide');
    $(window).scroll(function () {
        ($(window).scrollTop() > BACK_TO_TOP_ACTIVATION_HEIGHT) ? $('#backToTop').removeClass('hide') : $('#backToTop').addClass('hide');
    });
});
```
Does this make sense in the direction of simplifying? I believe the bootstrap classes are over-riding the inline styles with `!important`. The `hide` class is present below them in `min.css` file. Therefore, adding `hide` as a class to HTML element is not showing the desired effect. I load the file and on ready, add the `hide` class and then simply toggle this class with JS.

sushain97 replied:
Ah, this is a better approach. However, using a conditional expression like this is an anti-pattern. Use `toggleClass` for the cleanest code.

share-with-me replied:
Um, `toggleClass` with two arguments? Something like this:

```
$('#backToTop').addClass('hide');
$(window).scroll(function () {
    $('#backToTop').toggleClass('hide', $(window).scrollTop() < BACK_TO_TOP_ACTIVATION_HEIGHT);
});
```
